pxr/usd/pcp/layerStackRegistry.cpp

```diff
@@ -224,13 +232,21 @@ Pcp_LayerStackRegistry::GetAllLayerStacks() const
 // Private helper methods.
 
 void
-Pcp_LayerStackRegistry::_Remove(const PcpLayerStackIdentifier& identifier,
-                                const PcpLayerStack *layerStack)
+Pcp_LayerStackRegistry::_SetLayersAndRemove(
+    const PcpLayerStackIdentifier& identifier,
+    const PcpLayerStack *layerStack)
 {
+    tbb::queuing_rw_mutex::scoped_lock lock(_data->mutex, /*write=*/true);
+
     Pcp_LayerStackRegistryData::IdentifierToLayerStack::const_iterator i =
         _data->identifierToLayerStack.find(identifier);
-    if (TF_VERIFY(i != _data->identifierToLayerStack.end()) &&
-        TF_VERIFY(i->second.operator->() == layerStack)) {
+    // It's possible that layerStack has already been removed from the
+    // map if a FindOrCreate call intercedes between the moment when the
+    // layer stack's ref count drops to zero and the time the layer stack
+    // destructor is called (which is how we get into this method).
+    if (i != _data->identifierToLayerStack.end() &&
+        i->second.operator->() == layerStack) {
+        _SetLayers(layerStack);
```

Comment:
I think we should call _SetLayers outside of this if block. If we're in here, the layerStack is being destroyed so we'd want to make sure any mappings between it and other layers are cleaned up regardless of whether the identifierToLayerStack map has already been updated with a new entry for the given identifier. Does that seem right to you?

Reply:
Yes, I think you're right about this. My misunderstanding was that I thought the other mappings from the layer stack to layers and the layer stack to muted layers were also based on the layer stack identifier. But reading the code I see the other mappings are based on the layer stack pointer, so I agree they should always be cleaned up. Thanks!